File: src/main/java/xf/xflp/opt/construction/multitype/NContainerNTypeAddPacker.java
package xf.xflp.opt.construction.multitype;

import xf.xflp.base.XFLPModel;
import xf.xflp.base.container.Container;
import xf.xflp.base.item.Item;
import xf.xflp.exception.XFLPException;
import xf.xflp.opt.Packer;
import xf.xflp.opt.construction.strategy.Strategy;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Copyright (c) 2012-2022 Holger Schneider
 * All rights reserved.
 *
 * This source code is licensed under the MIT License (MIT) found in the
 * LICENSE file in the root directory of this source tree.
 *
 *
 * This packer puts the items in a sequence for each container type into single container.
 * Items will only be added to a container.
 *
 * @author hschneid
 *
 */
public class NContainerNTypeAddPacker implements Packer {

	@Override
	public void execute(XFLPModel model) throws XFLPException {

		Strategy strategy = model.getParameter().getPreferredPackingStrategy();
		MultiBinAddHeuristic heuristic = new MultiBinAddHeuristic(strategy, model.getStatusManager());

		List<Container> containers = new ArrayList<>();
		List<Item> unplannedItems = Arrays.asList(model.getItems());
		while(!unplannedItems.isEmpty()) {
			// Create one container per type
			List<Container> newContainers = getContainers(model);
			// Try to insert items in containers
			List<Item> restItems = heuristic.createLoadingPlan(unplannedItems, newContainers);

			containers.addAll(newContainers);
			// Rest containers will go into next round
			unplannedItems = restItems;
		}

		// Put result into model
		model.setContainers(containers.toArray(new Container[0]));
		model.setUnplannedItems(unplannedItems.toArray(new Item[0]));
	}

	private List<Container> getContainers(XFLPModel model) {
		return Arrays.stream(model.getContainerTypes())
				.map(Container::newInstance)
				.collect(Collectors.toList());
	}
}

File: src/xflp/opt/construction/multitype/n_container_n_type_add_packer.py
# Copyright (c) 2012-2025 Holger Schneider
# All rights reserved.
#
# This source code is licensed under the MIT License (MIT) found in the
# LICENSE file in the root directory of this source tree.

"""
This packer puts the items in a sequence for each container type into multiple containers.
Items will only be added to a container.

Author: hschneid
"""

from typing import List, TYPE_CHECKING

from xflp.opt.packer import Packer
from xflp.opt.construction.multitype.multi_bin_add_heuristic import MultiBinAddHeuristic

if TYPE_CHECKING:
    from xflp.base.xflp_model import XFLPModel
    from xflp.base.container.container import Container


class NContainerNTypeAddPacker(Packer):
    """Packer for multiple containers with multiple types (add only)."""

    def execute(self, model: 'XFLPModel') -> None:
        """
        Execute the packer.

        Args:
            model: The XFLP model

        Raises:
            XFLPException: If packing fails
        """
        strategy = model.get_parameter().get_preferred_packing_strategy()
        heuristic = MultiBinAddHeuristic(
            strategy,
            model.get_status_manager(),
            model.get_parameter()
        )

        containers: List['Container'] = []
        unplanned_items = list(model.get_items())

        while len(unplanned_items) > 0:
            # Create one container per type
            new_containers = self._get_containers(model)

            # Try to insert items in containers
            rest_items = heuristic.create_loading_plan(unplanned_items, new_containers)

            containers.extend(new_containers)
            # Rest containers will go into next round
            unplanned_items = rest_items

        # Put result into model
        model.set_containers(containers)
        model.set_unplanned_items(unplanned_items)

    def _get_containers(self, model: 'XFLPModel') -> List['Container']:
        """Create one container per type."""
        return [ct.new_instance() for ct in model.get_container_types()]
